### PR TITLE
Improve sql agent

### PR DESCRIPTION
**Pull request by ahuang11**
This was driving me nuts and took me a while to debug:

image

The shown SQL was valid, but internally, with `SQLLimit` transform, it was wrapping another layer of `SELECT *...` incorrectly, but the UI was not showing the final query...

1. Fixed with a regex to more robustly search for `SELECT ... FROM`; previously, it was doing `'select ' in`, but wasn't able to pick up `'select\n'`
2. added a jinja2 `examples` block so I can add example outputs, and users can also drop in examples, separately from instructions/context
3. changed the limit from 3 to 1000 because it was caching it, and it wasn't picking up enough enums to be useful for future cached queries
4. gave an example for expr_slug; sometimes it gave slugs with spaces "Top ..." which isn't a slug imo.
5. more checklist guidance for sqlagent

### CODE REVIEW

File: lumen/ai/prompts/SQLAgent/main.jinja2 [hunk 1]
```diff
@@ -26,8 +26,12 @@ Checklist
 - Quote column names to ensure they do not clash with valid identifiers.
 - If it's a date column (excluding individual year/month/day integers) date, cast to date using appropriate syntax, e.g.
 CAST or TO_DATE
-- Use only `{{ dialect }}` syntax
+- Use only `{{ dialect }}` SQL syntax
 - Try to pretty print the SQL output with newlines and indentation.
+- Specify data types explicitly to avoid type mismatches.
+- Handle NULL values using functions like COALESCE or IS NULL.
+- Use parameterized queries to prevent SQL injection attacks.
+- Use Common Table Expressions (CTEs) and subqueries to break down complex queries into manageable parts.
```
Comment (amaloney) [inline lumen/ai/prompts/SQLAgent/main.jinja2:34]: nice, CTEs are awesome

File: lumen/ai/prompts/SQLAgent/main.jinja2 [hunk 2]
```diff
@@ -40,3 +44,47 @@ identifiers.
...
+    COALESCE(TRY_CAST(regexp_extract(regexp_replace(athlete_medals, '[\n│]+', ' '), '(\d+)\s*S', 1) AS INTEGER), 0) AS silver_medals,
+    COALESCE(TRY_CAST(regexp_extract(regexp_replace(athlete_medals, '[\n│]+', ' '), '(\d+)\s*B', 1) AS INTEGER), 0) AS bronze_medals,
+    (
+        COALESCE(TRY_CAST(regexp_extract(regexp_replace(athlete_medals, '[\n│]+', ' '), '(\d+)\s*G', 1) AS INTEGER), 0) +
+        COALESCE(TRY_CAST(regexp_extract(regexp_replace(athlete_medals, '[\n│]+', ' '), '(\d+)\s*S', 1) AS INTEGER), 0) +
+        COALESCE(TRY_CAST(regexp_extract(regexp_replace(athlete_medals, '[\n│]+', ' '), '(\d+)\s*B', 1) AS INTEGER), 0)
+    ) AS total_medals
+FROM read_csv('olympic_athletes.csv')
+ORDER BY total_medals DESC
+LIMI5 1;
```
Comment (amaloney) [inline lumen/ai/prompts/SQLAgent/main.jinja2:67]: typo?

**Review by amaloney — CHANGES_REQUESTED** [review]
you need to fix the typo, but otherwise looks good

File: lumen/ai/prompts/SQLAgent/main.jinja2 (hunk 2 above)
Comment (ahuang11) [inline lumen/ai/prompts/SQLAgent/main.jinja2:67]:
```suggestion
LIMIT 1;
```

Comment (ahuang11) [reply lumen/ai/prompts/SQLAgent/main.jinja2]: Nice catch!

**Review by amaloney — APPROVED** [review]

**Review by philippjfr — APPROVED** [review]

File: lumen/ai/ui.py [hunk 3]
```diff
@@ -163,7 +163,7 @@ def _resolve_data(self, data: DataT | list[DataT] | None):
             source = DuckDBSource(
                 tables=tables, mirrors=mirrors,
                 uri=':memory:', initializers=initializers,
-                name='Provided'
+                name='ProvidedSource00000'
```
[the remark this answers is not in the document]
Comment (ahuang11) [reply lumen/ai/ui.py:166]: Strange that it's shown in this PR, but
> The LLM sometimes dropped the source, but by naming the Provided DuckDBSource like the other randomly named DuckDBSources, it seems to properly keep the name.



https://github.com/holoviz/lumen/pull/796

**Pull request merged**